### DISCUSSION TITLE
How are you dealing with heat ?  My RPI4 FLIRC case in the Create3 isn't cutting it. Any dood ideas for cooling ? Here's what I have so far.

### DISCUSSION THREAD

**Discussion by jerrynsr** [iRobotEducation/create3_docs#191] [Questions]
I had my Create3 w/RPI4 running ROS navigation missions in my house all morning long running a loop, iniitpose, undock, go to waypoint 1 while panting like a dog,  arrive at waypoint and bark,  sleep for 5 minutes, pant while navigating to waypoint 2, bark when arrive at waypoint, sleep for 5 minutes,   navigate to waypoint 3, do the same... sleep for 5 minutes,   go home (my work room),  pant for a bit,  dock, sleep for 15 minutes, and do the same thing over in a loop.  I had the loop set for 24.

Well it ran 12 times and then failed due to  CPU Temp reaching 84c.   I have the RPI 4 mounted inside a FLIRC case.  The FLIRC acts as one big heatsink.  The problem is the heat is contained in the cargo area of the bot.   I think a good fix will be a 3D printed cargo bay with a muffin fan mount.   I will try the mission again, this time I'll disable the DepthAI process and run with just LIDAR.

Better yet,  can iRobot mail me a Cargo Bay (The slider bay) ?  I can cut out a hole for two muffin fans and try that.

**Comment by raphaelcherney** [comment]
It would be cool to see a version of the cargo bay with a cooling fan! The STEP file for the Cargo Bay is available as part of the model [here](https://iroboteducation.github.io/create3_docs/hw/mechanical/#downloadable-3d-model). From there you can modify and print yourself if you have a printer or send it to a 3D printing service (e.g. https://craftcloud3d.com/). I'd love to see how it turns out!

**Reply by raphaelcherney** [reply]
I'm not sure what the top of your robot looks like, but another option is to mount the Pi to the faceplate instead of in the cargo bay which could simplify coming up with an active cooling solution.

**Reply by jerrynsr** [reply]
The Slamtec A-1 LIDAR and Oak D light camera is mounted there.

**Reply by jerrynsr** [reply]
No need for a 3D Printing Service.   I've got the Robo R1+ Printer.   I think I'll need to print this in ABS because PLA and PETG will warp due to the hear in the Cargo Bay.     

Thanks,

**Comment by shamlian** [comment]
We're hoping to have replacement faceplates and payload bays available for sale, soon.

**Reply by shamlian** [reply]
Soon is now! https://edu.irobot.com/shop/robot-accessories @jerrynsr

**Comment by jerrynsr** [comment]
It appears that the CPU load from processes running on the RPI4 for streaming video from the oak d lite to my desktop are a significant contributor for heating up the RPI4.   While running   create3_lidar and turtlebot4_navigation, the RPI4 is running at a  comfortable between 51C - 56C.   Here's the bash script I use to test. 
#!/bin/bash

init_bot(){
source ~/georgebot_ws/install/local_setup.bash
cd ~/georgebot_ws/robotest
ros2 launch create3_lidar sensors_launch.py &
ros2 launch turtlebot4_navigation nav_bringup.launch.py nav:=true slam:=off localization:=true map:=/home/jerryn/georgebot_ws/robotest/HomeMap.yaml &
# ~/depthai-python/examples/ColorCamera/rgb_preview.py &
sleep 20
}

robot_mission() {
./initpose
play -q ~/Sound/Dog/mixkit-medium-size-dog-walking-pant-58.wav
play -q ~/Sound/Dog/mixkit-medium-size-dog-walking-pant-58.wav &
./undock
play -q ~/Sound/Dog/mixkit-medium-size-dog-walking-pant-58.wav &
ros2 bag play ./jumpstart3
play -q ~/Sound/Dog/mixkit-medium-size-dog-walking-pant-58.wav &
./gotoden1
play -q ~/Sound/Dog/mixkit-medium-size-dog-walking-pant-58.wav &
sleep 60
play -q ~/Sound/Dog/mixkit-medium-size-dog-walking-pant-58.wav
play -q /home/jerryn/Sound/Dog/dog-2-low.wav &
./gotolivffp
play -q ~/Sound/Dog/mixkit-medium-size-dog-walking-pant-58.wav &
sleep 40
./gohome
play -q ~/Sound/Dog/mixkit-dog-whimper-sad-466.wav 
sleep 40
play -q ~/Sound/Dog/mixkit-medium-size-dog-walking-pant-58.wav &
./dock
play -q /home/jerryn/Sound/Dog/dogdrinkwater.wav
}

init_bot

numrun=15;

while [ $numrun -gt 0 ] ;
do
robot_mission
echo "Completed Mission Number: "$numrun
((numrun--))
sleep 900
# source ./unset_bot
done
killall ros2
killcam=`ps -ef |grep /home/jerryn/depthai-python/examples/ColorCamera/rgb_preview.py | grep -v color | cut -f7 -d" "`
killlidar=`ps -ef |grep /home/jerryn/georgebot_ws/install/rplidar_ros/lib/rplidar_ros/rplidar_composition | grep -v color | cut -f7 -d" "`
kill -9 $killcam
kill -9 $killlidar
ps -ef | grep ros/galactic | cut -f7 -d" " | xargs kill
rm -rf /tmp/tmp*
rm -rf /tmp/launch*
sleep 10
source ./unset_bot


If you are curious about my goto* scripts.. here's an example:

::::::::::::::
gotoden1
::::::::::::::
#!/bin/bash
ros2 topic pub -1 /goal_pose geometry_msgs/msg/PoseStamped '{ header: {stamp: {sec: 0, nanosec: 0}, frame_id: "map"}, pose: {position: {x: -4.145994186401367, y: -4.457064628601074, z: 0.0020704269409179688}, or
ientation: {w: 0.987692}}, }'

**Comment by ooijenvangilbert** [comment]
I am handeling the heat by installing the Rpi on top of the create 3, sort of like the turtebot setup.
As for the oak, I have setup a publisher and that doesn't seem to take up that many resources as your setup does.

**Reply by jerrynsr** [reply]
Cool.. publish the code!  You can post it here.

Does the heat dissipating from the RPI on top of the Create 3 cause issues with the LIDAR ?     The Laser on the SLAMTech A-1 is infrared so I figured heat may cause issues.

What Topics is your publisher broadcasting ?  Is it a reduced set ?     Do you see point cloud data in your room maps ?
I wanted to use both LIDAR and Depth Cam data for my maps and navigation.

**Reply by ooijenvangilbert** [reply]
I have an active cooler on the RPI, pushing airflow down, so I don't expect problems and there will be a plate between de RPI and lidar. Will finish that later this week. But good point , I will keep that in mind if the results are not what I expect.

At the moment I am still learning Ros, it's just a publisher, once I have a nice package I can put it up here.
Now it only publishes rgb images, every 0.1 seconds and scaled down to 300*300. I will work with those numbers to see where the sweet spot is. It was mainly to see where I was going when driving the create 3 over teleop. 

I see we have the same goals, maybe we should share a repo to help each other out? But as I said, still learning here.

**Reply by ooijenvangilbert** [reply]
I've added you to the repo, I just did a full map of my house with this script but will be optimizing it in the near future

**Reply by ooijenvangilbert** [reply]
i have a stereo point cloud example running as well, it's more that it is very laggy that hinders me than heat dissipated from the RPI. At the moment it's running on a 4b+ with only 2 gb of memory, I could test it with a 4b+ and 8 gb but first will see if I can optimize the code

**Comment by jerrynsr** [comment]
Ahh.. yeah.  I bet if you check the RPI it's running out of memory.
The 8GB RPI4 I have installed I bought before the shortage. You do not
want to run RVIZ on the RPI.  You want to use a remote ROS2 Node for
that.  I have a linux ROS2 desktop mode running for visualization.  
So I have three ROS2 Nodes running with my Create3 project.  The
Create3, RPI4, and my Linux Desktop.


My RPI4 also boots off and uses a m.2 SSD attached to one of the USB3
ports.

**Reply by ooijenvangilbert** [reply]
I can switch to the 8gb one, my config is sort of the same but on the RPI I am running ubuntu server without GUI, saves some gb's of memory there.

So my RVIZ is on a ubuntu installation on my laptop.
can an RPI boot from the SSD that way or do you still need a little card? 

One of the things that can also cause the laggy video stream is the frame rate coming from the camera and polling frequency of the publisher. if the frame.get() gets called with lower frequency than the cam delivers you see queuing and slowing of your feed. This is why I work with a second thread in the code, this one just races thru the frames and the publisher only picks the one when it's good and ready to do so.
Because the Oak does lots of work on the camera itself it's able to push our enormous frame rates.
